### ISSUE TITLE
Struggling with the dynamic keyboard key sizes (once again)

### ISSUE THREAD

**Issue by lsaranto** [JuliusSweetland/OptiKey#845]
I'm making a dynamic keyboard that shows the shortcut keys for the game Captain of Industry. I've laid out the keys, but some of the key labels look unnecessarily small (see image). How can I make the labels more readable? I tried to incorporate the tags that control the scaling, but it's not having an effect. I'm probably using them wrong. I've also attached the keyboard XML. 

![image](https://user-images.githubusercontent.com/52409168/175643811-46f5c72c-2672-4fc5-9efc-137ac8e337f7.png)

[CoI.txt](https://github.com/OptiKey/OptiKey/files/8979833/CoI.txt)

**Comment by tqphan** [comment]
@lsaranto Do any of these settings help?
https://i.imgur.com/wEoZt0t.png

**Comment by lsaranto** [comment]
> @lsaranto Do any of these settings help?

@tqphan Here's a crop from a screenshot of two keyboards next to each other. The bottom one is my first go for the same game. The top one is the one I showed in the first post. As you can see, OptiKey is putting a bigger margin on the key(s) in the new keyboard. Also, when going to smaller keys the fixed minimum margin is proportionally too much, IMO. I hope there was a way to use more space especially horizontally.

![image](https://user-images.githubusercontent.com/52409168/175806797-cad4483b-535d-4392-92bf-21ed5ab33dea.png)

**Comment by AdamRoden** [comment]
`AutoScaleToOneKeyWidth` and `SharedSizeGroup` are both supposed to be attributes. AutoScale has a default of true at both levels, so in order to override you need to set it to false in the keygroup and dynamickey. AutoScale will set your common keys to the same size based on the smallest key in the group of common keys. There are about 5 ways the identifies common keys so I won't get into it.

The bottom line is you need to set AutoScale to false twice because we weren't worried about this coming up as a problem when keygroups were introduced.

**Comment by AdamRoden** [comment]
`AutoScaleToOneKeyWidth` and `SharedSizeGroup` are both supposed to be attributes. AutoScale has a default of true at both the keygroup and dynamickey levels, so, if using keygroups, in order to override you need to set it to false in the keygroup and dynamickey. AutoScale will set your common keys to the same size based on the smallest key in the group of common keys. There are about 5 ways the code identifies common keys so I won't get into it.

The bottom line is you need to set AutoScale to false twice because we weren't worried about this coming up as a problem when keygroups were introduced.

**Comment by AdamRoden** [comment]
I think we could do something like this to make life easier.
![image](https://user-images.githubusercontent.com/45572061/176022376-3192c8bf-0d13-46e0-98f4-0bd1e67682e1.png)

**Comment by kmcnaught** [comment]
@AdamRoden thanks for the explanation. I've attempted to re-write the wiki information to make the situation clearer, let me know what you think...

**Comment by AdamRoden** [comment]
@kmcnaught You wrote accurately. I do think the code should be modified because by design an attribute at a DynamicKey level is always supposed to override the default or inherited value. Also, worthy of noting in the wiki is that the span used by AutoScaleToOneKeyWidth is the span of the smallest key. I think we could further improve AutoScaling by splitting this by blank, single letter, symbol, short text, and long text. I can probably PR those changes today if you agree with what I'm suggesting.

**Comment by kmcnaught** [comment]
I agree that the `AutoScaleToOneKey[Width/Height]` attribute on a key should override that of a `KeyGroup`. I would also say that if it is not defined on the key it should inherit from the `KeyGroup` - so you can turn the scaling off for an entire keygroup in one place. 

i.e. the `AutoScale*` properties should behave as follows:
- not specified - defaults to true
- specified on keygroup OR key - takes the value specified
- specified on keygroup AND key - the key's value takes precedence

I'm not sure what you mean by _splitting this by blank, single letter, symbol, short text, and long text_. Can you explain further please?

**Comment by lsaranto** [comment]
I wanted to edit the keyboard (I missed a crucial key) and ran into an issue. Since I had defined `Col` and `Row` in every key, adding or removing a key meant I had to correct every key's position. So I decided to let OptiKey lay out the keys and I removed every position attribute. What I got was a total mess. Unfortunately, I didn't save a screenshot.

According to the wiki, the positioning algorithm is this:

```
Position all content items that have both row and col defined
Position all remaining items in listed order
If an item has a row or column defined jump forward to that row or column and mark any/all skipped spaces as empty
```

It seems there's a step missing from the algorithm shown above. Optikey looks to be doing the keys with `Height` > 1 as a second phase. Only then does it the rest. So I had to place those keys in fixed positions to get what I wanted. It's still eight keys to correct if I need to modify the keyboard, but it's better than every key. Adding `Height="1"` to single high keys was not enough.

**Comment by AdamRoden** [comment]
@lsaranto I could try to refine the algorithm; just for fun. It does have caveats when Height > 1.

**Comment by AdamRoden** [comment]
@lsaranto Send me the two versions of your file.

**Comment by lsaranto** [comment]
@AdamRoden 
This one works:
[CoI_big.txt](https://github.com/OptiKey/OptiKey/files/9078639/CoI_big.txt)

This one I stripped all Row and Col attributes and it breaks:
[CoI_big_notworking.txt](https://github.com/OptiKey/OptiKey/files/9078642/CoI_big_notworking.txt)

**Comment by AdamRoden** [comment]
@lsaranto I rewrote the code yesterday and tested it today and the new code worked the first try. I feel good!

**Comment by lsaranto** [comment]
This fix is not yet included in a release?  Is it hard to compile the source code myself? What's the simplest way to do it?

**Comment by AdamRoden** [comment]
If you use Visual Studio you can easily clone a project and then compile the code for use. Don't consider this perfect instruction, but you could for example, clone https://github.com/AdamRoden/OptiKey.git to your local instance of Visual Studio. Then, make OptikeyPro the active solution and perform a Release Build. Once you've done that you'll get a new OptikeyPro.exe in your project's bin folder. You can run that exe or use it to replace the exe in your installed version of Optikey. The downfall is that the exe you build will not be signed so while it will work for most programs, it will not be registered my Windows as an accessibility app and will be denied certain permissions, for example use of Task Manager.

**Comment by lsaranto** [comment]
@AdamRoden Thank you, I managed to make the build. I did have to redo some settings in the management console (eg. sound) even after copying the config file. Also, somehow lost the default blue color on a key when a key is locked/pressed down. I tried to manually define it in the keyboard XML, but couldn't find the right shade. Do you know the name of it?

**Comment by lsaranto** [comment]
@AdamRoden Oh, I just noticed, that the link you gave was to your repo. Are there big differences between the original OptiKey repo and yours?

**Comment by AdamRoden** [comment]
@lsaranto Yes, my repo has some big changes such as a built-in keyboard designer, pop-up keys, and fixes to themes.

**Comment by kmcnaught** [comment]
@lsaranto fyi Adam's changes are lined up for bringing into Optikey once we've managed to do a bunch of testing. There will hopefully be a new stable release soon and then we can have a pre-release for everyone who's happy to test it all out.

(having said that, I don't know if there are also personal changes in Adam's repo that _aren't_ slated for a PR)
